device/src/efuse.rs

```diff
@@ -0,0 +1,198 @@
+// use core::num::NonZeroU8;
+use esp_hal::efuse::{self as hal_efuse, Efuse};
+use esp_hal::peripherals::EFUSE;
+use rand_chacha::rand_core::RngCore;
+use reed_solomon;
+
+pub struct EfuseController {
+    pub efuse: EFUSE,
+}
+
+impl EfuseController {
+    pub fn new(efuse: EFUSE) -> Self {
+        Self { efuse }
+    }
+
+    pub fn init_key(&self, key_number: u8, rng: &mut impl RngCore) -> Result<(), EfuseError> {
+        let efuse_field = match key_number {
+            0 => hal_efuse::KEY_PURPOSE_0,
+            1 => hal_efuse::KEY_PURPOSE_1,
+            2 => hal_efuse::KEY_PURPOSE_2,
+            3 => hal_efuse::KEY_PURPOSE_3,
+            4 => hal_efuse::KEY_PURPOSE_4,
+            5 => hal_efuse::KEY_PURPOSE_5,
+            _ => return Err(EfuseError::EfuseError),
+        };
+        let key: u8 = Efuse::read_field_le(efuse_field);
+        // Check if there's an existing key so we don't accidentally overwrite it
+        if key == 0 {
+            let mut buff = [0x00_u8; 32];
+            rng.fill_bytes(&mut buff);
+
+            unsafe {
+                self.write_block(&buff, key_number + 4)?;
+                self.write_key_purpose(key_number, KeyPurpose::HmacUpstream)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// # Safety
+    unsafe fn write_key_purpose(
+        &self,
+        key_number: u8,
+        key_purpose: KeyPurpose,
+    ) -> Result<(), EfuseError> {
+        let mut buff = [0x00u8; 32];
+
+        let kp = key_purpose as u8;
+        match key_number {
+            0 => buff[11] = kp,
+            1 => buff[11] = kp << 4,
+            2 => buff[12] = kp,
+            3 => buff[12] = kp << 4,
+            4 => buff[13] = kp,
+            5 => buff[13] = kp << 4,
+            _ => return Err(EfuseError::EfuseError),
+        }
+
+        // We bundle every config in Block 0 to minimize write operations
+        // Todo: Write key purpose and rw flags for multiple keys
+        // Disable write to key block
+        let mut write_disable = 0x01_u32 << (23 + key_number);
+        // Disable write to key purpose
+        write_disable += 0x01_u32 << (8 + key_number);
+        buff[0..4].copy_from_slice(&write_disable.to_le_bytes());
+
+        #[cfg(feature = "efuse_protect")]
+        self.set_read_protect(key_number, &mut buff);
```

Comment:
:wrench:  Have a bool to this his function which decides whether it should be read protected. The bool can be set in the main function by the feature flag. like:

```
let do_read_protect = #[cfg(feature = "efuse_protect")];
```

The name of the feature should be `read_protect_hmac_key`.